File: java/1139Largest1-BorderedSquare.java
class Solution {
    public int largest1BorderedSquare(int[][] grid) {
        int[][][] table = new int[grid.length][grid[0].length][2];
        for (int i = 0; i < grid.length; ++i) {
            for (int j = 0; j < grid[0].length; ++j) {
                if (grid[i][j] == 1) {
                    table[i][j][0] = 1;
                    table[i][j][1] = 1;
                    if (j > 0) {
                        table[i][j][0] += table[i][j-1][0];
                    }
                    if (i > 0) {
                        table[i][j][1] += table[i-1][j][1];
                    }
                }
            }
        }

        int max = 0;
        for (int i = 0; i < grid.length; ++i) {
            for (int j = 0; j < grid[0].length; ++j) {
                if (grid[i][j] == 1) {
                    int min = Math.min(table[i][j][0], table[i][j][1]);
                    while (min > max) {
                        if (table[i][j - min + 1][1] >= min && table[i  - min + 1][j][0] >= min) {
                            max = min;
                        }

                        --min;
                    }
                }
            }
        }

        return max * max;
    }
}

File: cplusplus/1139Largest1-BorderedSquare.cpp
class Solution {
public:
    int largest1BorderedSquare(vector<vector<int>>& grid) {
        vector<vector<int>> horizon(grid.size(), vector<int>(grid[0].size(), 0));
        vector<vector<int>> vertical(grid.size(), vector<int>(grid[0].size(), 0));

        int max = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == 1) {
                    vertical[i][j] = 1;
                    horizon[i][j] = 1;
                    if (j > 0) {
                        horizon[i][j] += horizon[i][j-1];
                    }
                    if (i > 0) {
                        vertical[i][j] += vertical[i-1][j];
                    }
                    int len = horizon[i][j];
                    if (len > vertical[i][j]) {
                        len = vertical[i][j];
                    }

                    while (len > max) {
                        if (vertical[i][j - len + 1] >= len && horizon[i - len + 1][j] >= len) {
                            max = len;
                        }
                        --len;
                    }
                }
            }
        }

        return max * max;
    }
};

=================================================================================

class Solution {
public:
    int largest1BorderedSquare(vector<vector<int>>& grid) {
        vector<vector<int>> horizon(grid.size(), vector<int>(grid[0].size(), 0));
        vector<vector<int>> vertical(grid.size(), vector<int>(grid[0].size(), 0));

        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == 1) {
                    horizon[i][j] = 1;
                    vertical[i][j] = 1;
                    if (i > 0) {
                        vertical[i][j] += vertical[i-1][j];
                    }
                    if (j > 0) {
                        horizon[i][j] += horizon[i][j-1];
                    }
                }
            }
        }

        int max = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == 1) {
                    int len = horizon[i][j];
                    if (len > vertical[i][j]) {
                        len = vertical[i][j];
                    }

                    while (len > max) {
                        if (vertical[i][j - len + 1] >= len && horizon[i - len + 1][j] >= len) {
                            max = len;
                        }
                        --len;
                    }
                }
            }
        }

        return max * max;
    }
};

